hledger-lib/Hledger/Data/Amount.hs

```diff
@@ -580,56 +579,64 @@ canonicaliseAmount styles a@Amount{acommodity=c, astyle=s} = a{astyle=s'}
 
 instance Semigroup MixedAmount where
   (<>) = maPlus
+  sconcat = maSum
+  stimes n = multiplyMixedAmount (fromIntegral n)
 
 instance Monoid MixedAmount where
   mempty = nullmixedamt
+  mconcat = maSum
 #if !(MIN_VERSION_base(4,11,0))
   mappend = (<>)
 #endif
 
 instance Num MixedAmount where
-  fromInteger i = Mixed [fromInteger i]
-  negate = maNegate
-  (+)    = maPlus
-  (*)    = error' "error, mixed amounts do not support multiplication" -- PARTIAL:
-  abs    = error' "error, mixed amounts do not support abs"
-  signum = error' "error, mixed amounts do not support signum"
-
--- | Get a mixed amount's component amounts.
-amounts :: MixedAmount -> [Amount]
-amounts (Mixed as) = as
+    fromInteger = mixedAmount . fromInteger
+    negate = maNegate
+    (+)    = maPlus
+    (*)    = error' "error, mixed amounts do not support multiplication" -- PARTIAL:
+    abs    = error' "error, mixed amounts do not support abs"
+    signum = error' "error, mixed amounts do not support signum"
+
+-- | Find the key used for an Amount within MixedAmount.
+amountKey :: Amount -> MixedAmountKey
+amountKey Amount{acommodity=c,aprice=p} = (c, priceKey <$> p)
+  where
+    priceKey (UnitPrice  x) = (acommodity x, Just $ aquantity x)
+    priceKey (TotalPrice x) = (acommodity x, Nothing)
 
 -- | The empty mixed amount.
 nullmixedamt :: MixedAmount
-nullmixedamt = Mixed []
+nullmixedamt = Mixed mempty
 
 -- | A temporary value for parsed transactions which had no amount specified.
 missingmixedamt :: MixedAmount
 missingmixedamt = mixedAmount missingamt
 
--- | Convert amounts in various commodities into a normalised MixedAmount.
-mixed :: [Amount] -> MixedAmount
-mixed = normaliseMixedAmount . Mixed
+-- | Convert amounts in various commodities into a mixed amount.
+mixed :: Foldable t => t Amount -> MixedAmount
+mixed = maAddAmounts nullmixedamt
+{-# SPECIALISE mixed :: [Amount] -> MixedAmount #-}
```

Comment:
Adding these SPECIALISE lines causes GHC to generate an extra version that is optimised for lists. Did this make a noticeable difference ? (And just for the record, with which GHC version ?)

Reply:
I did not check if there was any performance improvement. Previously `mixed` worked only on `[Amount]`, and I added the specialise pragma to make sure that this most-common case was always optimised for.

Reply:
I tend to feel we shouldn't bother adding such pragmas unless/until we know they give a significant benefit, since they tend to be unpredictable across GHC versions and they cost extra maintenance and code-reading effort.

Reply:
Got it.